### PR TITLE
[DO-NOT-MERGE][SPARK-23710] Upgrade built-in Hive to 2.3.4(Without hive-thriftserver)

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

This is the first PR. Just to make it easier to review the changes except `hive-thriftserver` module.
Complete changes please go:

## How was this patch tested?

unit tests and manual tests

### CODE REVIEW

File: core/pom.xml [hunk 1]
```diff
@@ -391,16 +391,42 @@
      placed in the "provided" scope, rather than the "compile" scope, and NoClassDefFoundError exceptions are handled
      when the user has not explicitly compiled with the Hive module.
     -->
+    <dependency>
```
Comment (srowen) [inline core/pom.xml]: Do these dependencies have to come into core?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala [hunk 2]
```diff
@@ -253,6 +253,12 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
     dir
   }
 
+  private def isSubDir(p1: Path, p2: Path, fs: FileSystem): Boolean = {
+    val path1 = fs.makeQualified(p1).toString
+    val path2 = fs.makeQualified(p2).toString
+    if (path1.startsWith(path2)) true else false
```
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala]: No need for `if (...) true else false`. Just return the value of the predicate

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala [hunk 3]
```diff
@@ -197,6 +201,13 @@ class OrcFileFormat extends FileFormat with DataSourceRegister with Serializable
 
     case _ => false
   }
+
+  def toKryo(sarg: SearchArgument): String = {
```
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: private?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala [hunk 4]
```diff
@@ -82,6 +83,39 @@ private[orc] object OrcFilters extends Logging {
...
+
+  /**
+   * Cast literal values for filters.
+   *
+   * We need to cast to long because ORC raises exceptions
+   * at 'checkLiteralType' of SearchArgumentImpl.java.
+   */
+  private def castLiteralValue(value: Any, dataType: DataType): Any = dataType match {
+    case ByteType | ShortType | IntegerType | LongType =>
+      value.asInstanceOf[Number].longValue
```
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: Trivial, but I'd use () or not consistently between here and line 110

**Review by srowen — COMMENTED** [review]
Is it still necessary to have the hive-contrib-2.3.4.jar etc in the test resources, to make the tests work?

Could you summarize the compatibility changes in the "Docs text" field of the JIRA and add the release-notes tag? that would help everyone understand the implications of the change.

But yeah this seems like something we have to do.

File: core/pom.xml (hunk 1 above)
Comment (rdblue) [reply core/pom.xml]: I think this is okay because it uses the "core" version of hive-exec. The version without a classifier shades everything, so I think the intent here is to manage Hive's transitive dependencies in Spark instead of in a fork of Hive.

Comment (srowen) [reply core/pom.xml]: Oh right it's `provided` anyway

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala [hunk 5]
```diff
@@ -192,6 +192,7 @@ private[hive] class IsolatedClientLoader(
     (name.startsWith("com.google") && !name.startsWith("com.google.cloud")) ||
     name.startsWith("java.lang.") ||
     name.startsWith("java.net") ||
+    name.startsWith("org.apache.derby.") ||
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: This breaks Livy, as @HyukjinKwon pointed out. https://github.com/apache/spark/pull/20944#issuecomment-436530497

File: sql/core/pom.xml [hunk 6]
```diff
@@ -86,15 +86,17 @@
       <scope>test</scope>
     </dependency>
 
+    <dependency>
+      <groupId>org.apache.hive</groupId>
```
Comment (gatorsmile) [inline sql/core/pom.xml]: We should not add the dependence to hive.

File: sql/core/pom.xml [hunk 7]
```diff
@@ -86,15 +86,17 @@
...
+      <artifactId>hive-storage-api</artifactId>
+    </dependency>
     <dependency>
       <groupId>org.apache.orc</groupId>
       <artifactId>orc-core</artifactId>
-      <classifier>${orc.classifier}</classifier>
```
Comment (gatorsmile) [inline sql/core/pom.xml]: Our native orc upgrade should be independent on Hive ORC reader, as @dongjoon-hyun pointed out.

**Review by gatorsmile — CHANGES_REQUESTED** [review]
I really doubt the value of the upgrade of Hive execution JARs based on the risk it brings.

**Comment by srowen** [comment]
@gatorsmile @wangyum I agree with the risks here; one important consideration too is that I don't think Hive 1.x will work at all with Java 9+. Still investigating but it looks like Hive 2.x might. That doesn't mean we have to merge this but may come up again as a tough call to make for 3.0.

**Comment by wangyum** [comment]
@srowen We plan to upgrade the built-in Hive to 2.3.4 for hadoop-3.1. hadoop-2.7 still uses 1.2.1. more details: https://issues.apache.org/jira/browse/SPARK-23710

**Comment by srowen** [comment]
That could make sense, to only publish / test with Java 11 with Hadoop 3.x and Hive 2.x+
